Add disjoint bindImageMemory VUID

sfricke-samsung opened the pull request:
This adds
VUID-VkBindImageMemoryInfo-pNext-01619
VUID-VkBindImageMemoryInfo-pNext-01620
VUID-VkBindImageMemoryInfo-pNext-01621

Which already were checked for non-disjoint VUID, this adds support for disjoint VUID.

This required some additional state tracking logic to be added as well as removing unused arguments getting passed around

Also adds
VUID-VkBindImageMemoryInfo-pNext-01615
VUID-VkBindImageMemoryInfo-pNext-01616
VUID-VkBindImageMemoryInfo-pNext-01617
Which is just the `device_extensions.vk_khr_sampler_ycbcr_conversion` addition of 3 other VUID

jeremy-lunarg commented:
Seeing a failure on an android device:

[ RUN      ] VkLayerTest.BindInvalidMemory
Unexpected: [ VUID-VkBindImageMemoryInfo-pNext-01619 ] Object: 0x130000000013 (Type = 8) | vkBindImageMemory2KHR() pBindInfos[0](): MemoryRequirements->memoryTypeBits (0x0) for this object type are not compatible with the memory type (0x0) of VkDeviceMemory 0x130000000013[]. The Vulkan spec states: If the pNext chain includes a VkBindImagePlaneMemoryInfo structure, memory must have been allocated using one of the memory types allowed in the memoryTypeBits member of the VkMemoryRequirements structure returned from a call to vkGetImageMemoryRequirements2 with image and where VkBindImagePlaneMemoryInfo::planeAspect corresponds to the VkImagePlaneMemoryRequirementsInfo::planeAspect in the VkImageMemoryRequirementsInfo2 structure's pNext chain (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkBindImageMemoryInfo-pNext-01619)
Other error messages logged for this test were:
     [ VUID-VkBindImageMemoryInfo-pNext-01619 ] Object: 0x130000000013 (Type = 8) | vkBindImageMemory2KHR() pBindInfos[0](): MemoryRequirements->memoryTypeBits (0x0) for this object type are not compatible with the memory type (0x0) of VkDeviceMemory 0x130000000013[]. The Vulkan spec states: If the pNext chain includes a VkBindImagePlaneMemoryInfo structure, memory must have been allocated using one of the memory types allowed in the memoryTypeBits member of the VkMemoryRequirements structure returned from a call to vkGetImageMemoryRequirements2 with image and where VkBindImagePlaneMemoryInfo::planeAspect corresponds to the VkImagePlaneMemoryRequirementsInfo::planeAspect in the VkImageMemoryRequirementsInfo2 structure's pNext chain (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkBindImageMemoryInfo-pNext-01619)
jni/../../tests/vkrenderframework.cpp:162: Failure
Failed
Did not receive expected error 'VUID-VkBindImageMemoryInfo-pNext-01620'
Unexpected: [ VUID-VkBindImageMemoryInfo-pNext-01619 ] Object: 0x140000000014 (Type = 8) | vkBindImageMemory2KHR() pBindInfos[0](): MemoryRequirements->memoryTypeBits (0x0) for this object type are not compatible with the memory type (0x0) of VkDeviceMemory 0x140000000014[]. The Vulkan spec states: If the pNext chain includes a VkBindImagePlaneMemoryInfo structure, memory must have been allocated using one of the memory types allowed in the memoryTypeBits member of the VkMemoryRequirements structure returned from a call to vkGetImageMemoryRequirements2 with image and where VkBindImagePlaneMemoryInfo::planeAspect corresponds to the VkImagePlaneMemoryRequirementsInfo::planeAspect in the VkImageMemoryRequirementsInfo2 structure's pNext chain (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkBindImageMemoryInfo-pNext-01619)
Other error messages logged for this test were:
     [ VUID-VkBindImageMemoryInfo-pNext-01619 ] Object: 0x140000000014 (Type = 8) | vkBindImageMemory2KHR() pBindInfos[0](): MemoryRequirements->memoryTypeBits (0x0) for this object type are not compatible with the memory type (0x0) of VkDeviceMemory 0x140000000014[]. The Vulkan spec states: If the pNext chain includes a VkBindImagePlaneMemoryInfo structure, memory must have been allocated using one of the memory types allowed in the memoryTypeBits member of the VkMemoryRequirements structure returned from a call to vkGetImageMemoryRequirements2 with image and where VkBindImagePlaneMemoryInfo::planeAspect corresponds to the VkImagePlaneMemoryRequirementsInfo::planeAspect in the VkImageMemoryRequirementsInfo2 structure's pNext chain (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkBindImageMemoryInfo-pNext-01619)
jni/../../tests/vkrenderframework.cpp:162: Failure
Failed
Did not receive expected error 'VUID-VkBindImageMemoryInfo-pNext-01620'
[  FAILED  ] VkLayerTest.BindInvalidMemory (224 ms)

sfricke-samsung commented:
@jeremy-lunarg do you have a dev-sim json file of that device so I could triage it. Will also take another look overall as well

sfricke-samsung commented:
@jeremy-lunarg ping... looking at error seems the driver is returning `0` to `vkGetImageMemoryRequirements2` which not sure if a driver bug or something I need to catch in the layer... I have a few set of devices so hoping you can give me details on what device is failing and hoping to reproduce

jeremy-lunarg commented:
@sfricke-samsung sorry for the delay and thanks for that bit of info. Things have been a bit hectic. I've asked @mark-lunarg to help with this one. His android-fu is stronger than mine.

jeremy-lunarg commented:
@sfricke-samsung the device is a galaxy S10

sfricke-samsung commented:
@jeremy-lunarg thanks no problem, just me giving a friendly poke, nothing super urgent (also going to assume you guys have US version which would have Adreno driver in it)

sfricke-samsung commented:
@jeremy-lunarg  So the Galaxy S9 with a 1.1.87 driver doesn't reproduce the issue, but looking at the error 

> MemoryRequirements->memoryTypeBits (0x0) for this object type are not compatible with the memory type (0x0) of VkDeviceMemory

It seems that `memoryTypeBits` was returned zero and looking at the spec

> The implementation guarantees certain properties about the memory requirements returned by vkGetBufferMemoryRequirements and vkGetImageMemoryRequirements:
>
> -  The memoryTypeBits member always contains at least one bit set.

So I am going to file a spec issue internally to see if it should also include `vkGetImageMemoryRequirements2` or `vkGetBindMemoryRequirements2` as well

The layer code will still work regardless, but for the 2 tests for VUID 01620 I have added a check for `memoryTypeBits` so when you re-run on CI it shouldn't false positive. (I also rebased as well)

jeremy-lunarg commented:
I'm still seeing the errors with your new patch. I'll take another look. I'd like to blacklist this test on the device, but there's a LOT of sub-tests in this test. It'd be nice if they were more atomic.

sfricke-samsung commented:
agree, this test got pretty bloated with this change... Well a quick thing to try is to just have `CoreChecks::ValidateMemoryTypes` return `false` if `memoryTypeBits` is zero as that is not allowed in the spec and if it returning zero I found out thats a driver bug, but still is valid in terms of validation layers

jeremy-lunarg commented:
I don't think the problem is in the driver. `vkGetImageMemoryRequirements2` is returning non-zero `memoryTypeBits` but `ValidateMemoryTypes` is getting zero `memoryTypeBits`.

sfricke-samsung commented on layers/state_tracker.cpp:
```diff
@@ -2133,28 +2153,48 @@ void ValidationStateTracker::PostCallRecordGetBufferMemoryRequirements2KHR(VkDev
+                                                                  VkMemoryRequirements *pMemoryRequirements) {
+    const VkImagePlaneMemoryRequirementsInfo *plane_info =
+        (pInfo == nullptr) ? nullptr : lvl_find_in_chain<VkImagePlaneMemoryRequirementsInfo>(pInfo->pNext);
+    // TODO does the VkMemoryRequirements need to be saved here if PostCallRecordCreateImage tracks it regardless
```
(side note to self: I need to update this type function to `Requirements` lol)

jeremy-lunarg commented:
I might take a stab at refactoring this test into smaller tests. It's making my head spin.

sfricke-samsung commented:
If you want, I can do that today (I need to rebase anyways so let me know)

Sorry for cramming it into that other tests... its hard to know when to make a new test and when not to reinvent a test's wheels

jeremy-lunarg commented:
That'd be great! Personally, I am a fan of the smallest possible unit test, but no one agrees with me. :smile: 

At the least, it'd be nice to split up the highest-level unnamed scopes, but there's many ways to slice it so it's your call.

sfricke-samsung commented:
So really sorry @jeremy-lunarg for even make you attempt to look at that old test... it was quite the mess... I decided to add another commit and rightfully moves the YCbCr tests that were there to its own test on top of creating a fresh new test ONLY for the VUIDs added.

I also found out that I was breaking an undone VU where you have to bind all the disjoint image planes at the same time... I have plans to add `VUID-vkBindImageMemory2-pBindInfos-02858` soon, but technically the error might have been Undefined Behavior... hopefully this just works, if not let me know

jeremy-lunarg commented:
@sfricke-samsung Don't be sorry! Thanks for all the work you've be doing. I'm re-running these patches through CI. I'll report back in a bit.

jeremy-lunarg commented:
@sfricke-samsung android is clean now but seeing some errors on nvidia and amd:

NVIDIA:
[ RUN      ] VkLayerTest.DedicatedAllocation
Unexpected: Validation Error: [ VUID-VkMemoryDedicatedAllocateInfo-image-01433 ] Object 0: handle = 0x6c429f0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xd47f82b6 | Allocation Size (2048) needs to be equal to VkImage VkImage 0x70000000007[] VkMemoryRequirements::size (0) The Vulkan spec states: If image is not VK_NULL_HANDLE, VkMemoryAllocateInfo::allocationSize must equal the VkMemoryRequirements::size of the image (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkMemoryDedicatedAllocateInfo-image-01433)
/home/lunarg/.jenkins/w8/64-bit/Debug/tests/vkrenderframework.cpp:172: Failure
Failed
Received unexpected error(s).
[  FAILED  ] VkLayerTest.DedicatedAllocation (368 ms)

AMD:
[ RUN      ] VkLayerTest.BindInvalidMemory2Disjoint
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
VUID-vkDestroyDevice-device-00378(ERROR / SPEC): msgNum: 1901072314 - Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x2a60bed1c60, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0x5d11410000000004, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x2a60bed1c60[], VkDeviceMemory 0x5d11410000000004[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyDevice-device-00378)
    Objects: 2
        [0] 0x2a60bed1c60, type: 3, name: NULL
        [1] 0x5d11410000000004, type: 8, name: NULL
VUID-vkDestroyDevice-device-00378(ERROR / SPEC): msgNum: 1901072314 - Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x2a60bed1c60, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0x983e60000000003, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x2a60bed1c60[], VkImage 0x983e60000000003[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyDevice-device-00378)
    Objects: 2
        [0] 0x2a60bed1c60, type: 3, name: NULL
        [1] 0x983e60000000003, type: 10, name: NULL
[  FAILED  ] VkLayerTest.BindInvalidMemory2Disjoint (696 ms)

sfricke-samsung commented:
well glad as easier to debug these... think I know what I did, will try and clean these quick right now

sfricke-samsung commented:
So found the issue with the NVIDIA card (I reworked the case around 01433 to be correct)

The AMD one I will try later when I build for my Windows machine (which has an AMD card), but looks at first glance it failed/crashed on one of the last test in BindInvalidImage2Disjoint and because it crashed out it never hit the clean up.

sfricke-samsung commented:
errr... can't reproduce AMD driver issue...

jeremy-lunarg commented:
grr... great... I think the AMD issue is intermittent so I'm not even sure it's a problem with this patch.

sfricke-samsung commented on tests/vklayertests_buffer_image_memory_sampler.cpp:
```diff
@@ -1464,14 +1448,202 @@ TEST_F(VkLayerTest, BindInvalidMemory) {
+    PFN_vkBindImageMemory2KHR vkBindImageMemory2Function = nullptr;
+    PFN_vkGetImageMemoryRequirements2KHR vkGetImageMemoryRequirements2Function = nullptr;
+
+    if (mp_extensions) {
```
oooo crap...
I mixed up `VK_KHR_get_memory_requirements2` and `VK_KHR_bind_memory2`... 
will fix, I also really really hated this stupid ifdef in the spec with all these extensions

jeremy-lunarg replied:
I feel your pain! :wink:

sfricke-samsung replied:
So the issue is when I use --devsim with mockICD it enabled ALL the extensions... I realize devsim only catches physical device properties so really the issue is MockICD (to my knowledge) not having an easy way to toggle extension support without rebuilding it... not sure how useful it would be to allow devsim layer to also intercept device extension queries as well... will raise question in the appropriate repo </rant>... regardless, really really really really hope this last patch is the golden patch that works everywhere!

jeremy-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2828,24 +2828,25 @@ bool CoreChecks::PreCallValidateAllocateMemory(VkDevice device, const VkMemoryAl
-                                 "VK_IMAGE_CREATE_SPARSE_BINDING_BIT",
-                                 report_data->FormatHandle(dedicated_allocate_info->image).c_str());
-            }
-            if ((device_extensions.vk_khr_sampler_ycbcr_conversion) &&
```
Is this going to be safe on vulkan 1.0?

jeremy-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2828,24 +2828,25 @@ bool CoreChecks::PreCallValidateAllocateMemory(VkDevice device, const VkMemoryAl
-                                 "VK_IMAGE_CREATE_SPARSE_BINDING_BIT",
-                                 report_data->FormatHandle(dedicated_allocate_info->image).c_str());
-            }
-            if ((device_extensions.vk_khr_sampler_ycbcr_conversion) &&
```
Why is the extension check removed?

sfricke-samsung replied:
`VK_IMAGE_CREATE_DISJOINT_BIT` was introduced in `vk_khr_sampler_ycbcr_conversion` so to my knowledge, this is redundant as if the flag had `VK_IMAGE_CREATE_DISJOINT_BIT` there should be some parameter checking to flag that down

jeremy-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -2147,28 +2167,48 @@ void ValidationStateTracker::PostCallRecordGetBufferMemoryRequirements2KHR(VkDev
+                                                                   VkMemoryRequirements *pMemoryRequirements) {
+    const VkImagePlaneMemoryRequirementsInfo *plane_info =
+        (pInfo == nullptr) ? nullptr : lvl_find_in_chain<VkImagePlaneMemoryRequirementsInfo>(pInfo->pNext);
+    // TODO does the VkMemoryRequirements need to be saved here if PostCallRecordCreateImage tracks it regardless
```
@mark-lunarg Can you answer this TODO?

sfricke-samsung replied:
So something I didn't want to tackle in the PR, but after this gets merged, I wanted to look into if we just don't need to save things from `GetImageMemoryRequirements` calls because we already capture the data when the image is created

It is legal to not call `VkGetImageMemoryRequirements` and thats why the layer needs to take precaution and do it at CreateImage time... with that, this shouldn't be needed as it seems to just be tracking stuff twice

Note this PR was to add disjoint tracking, another PR would be good to Remove this IMO

jeremy-lunarg replied:
I see. I had assumed that `VkGetImageMemoryRequirements` calls are required. Thanks for clearing that up.

sfricke-samsung replied:
ya, it is checked and then in best practices it yells at you if you didn't

The one case that has a nasty edge case is when importing External Android Hardware Buffers as you can't call `VkGetImageMemoryRequirements` on it until the AHB memory is bound because the driver won't actually know the size of the AHB until you map it with a BindCall

That and if apps feel like playing with knives and wing it without calling it 😎 lol

jeremy-lunarg commented:
:beers:
